### ISSUE TITLE
Re-implment balance function with IMU

### ISSUE THREAD

**Issue by hoverboardhavoc** [RoboDurden/Hoverboard-Firmware-Hack-Gen2.x#92]
It would be interesting to make a hoverboard balance again with this firmware. I think the chip marked with 16A 3C6A C3C might be an IMU. I've Googled and I can't find any information. How can I work out how to interface with the chip? I assume it must be a SPI or possibly I2C bus. I think that [extracting the original firmware](https://www.usenix.org/system/files/woot20-paper-obermaier.pdf) is tricky, at least for me. Maybe there is a way to run the original firmware and spy on the signals sent from the micro controller to the chip?

![P_20240702_230651](https://github.com/user-attachments/assets/a31f243d-b6a8-4ead-9bab-9a1cfa783f83)

**Comment by RoboDurden** [comment]
if you want this rather stupid self balancing toy, then buy a new one..
You could try to run the arduino  i2c scanner  code to get the address this little chips responds to.
That might give you a clue to what type of imu it is.
But IMU is not at all the scope of this firmware.

**Comment by I-hate-2FA** [comment]
the imu is connected with i2c, but its better to use a new imu and connect it to your esp

**Comment by hoverboardhavoc** [comment]
Thanks for the responses, I have a few of these Classy Walks to play with. 

I agree that just recreating the existing functionality is boring, but it might be fun to turn it into a two wheeled remote controlled vehicle.

**Comment by I-hate-2FA** [comment]
just put a imu on esp and write balance code on it, 100ms latency should be ok to balance if the mechanical design is not terrible

**Comment by I-hate-2FA** [comment]
but dont share your project with too much unnecessary detail on this repo, its for your own good

**Comment by hoverboardhavoc** [comment]
ok thanks for the help. I'll close this one

**Comment by hoverboardhavoc** [comment]
I made a start on this https://github.com/hoverboardhavoc/HoverboardImu/

**Comment by RoboDurden** [comment]
Yes nice 🙂 
Good start with a detailed read.me.

Maybe you want to look at all the different layouts here to find some more boards using this IMU.

**Comment by RoboDurden** [comment]
Will your code also compile with the Keil IDE ?

**Comment by RoboDurden** [comment]
no it is a normal ebike hub motor with 3phases and hall cables. Should work with a gen2.1 board.
I do not have the happiness to port the code myself into Gen2.x
And when i look at https://github.com/hoverboardhavoc/hoverboard-gen2.1-hack-GD-imu/tree/main/Src is see too many new files.
If you can merge all the i2c and imu code into one mpu6050.c and mpu6050.h file, that also contains the setup code with only two functions to be called from main() imuSetup() and imuUpdate(), then it would be easy for me to add to Gen2.x

**Comment by hoverboardhavoc** [comment]
Another update: The code on the [branch](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/compare/main...hoverboardhavoc:Hoverboard-Firmware-Hack-Gen2.x-GD32:self_balance) works but is slow. I'm considering removing the mpu6050 library pasted from the https://github.com/EFeru/hoverboard-sideboard-hack-GD project and going for something much simpler that does not use interrupts since my IMU does not support DMP.

**Comment by hoverboardhavoc** [comment]
https://github.com/user-attachments/assets/e69cd9da-d3ae-4ed9-b33d-d05f37e7b797

**Comment by RoboDurden** [comment]
That video really looks great.
If you can isolate the imu code to just one imu.c / imu.h pair, i really would like to add the code gen2.x

@DDoru If you have such boards to spare, i would send your 5 Euro via paypal for shipping to Germany. Maybe also 10 Euro after i have searched my stock of hoverboards and found now gen2 with such an imu.

Maybe i could not only detect braking but also pedaling on an ebike as a periodic little rotation alongside the driving direction..

**Comment by hoverboardhavoc** [comment]
I want 4 files 😱

generic i2c functions: i2c.h / c
mpu6050: mp6050.h / c
balance logic / pid controller: balance.h / c
Attitude And Heading Reference System (AHRS) algorithm (pasted code) MadgwickAHRS.c /h

You would only need the first two @RoboDurden if you want the raw data from the IMU. Obviously they will all be enabled/disabled with `#ifdef`s

**Comment by hoverboardhavoc** [comment]
Join us on Platform IO @RoboDurden 🤗

**Comment by RoboDurden** [comment]
I hope that i will get BLDC_SINE to work for 32f103 today.
So i would not mind f103 if your imu code runs on it.
here my email for shipping address and your paypal account: roland@robodurden.de
Where are you from ? Should be inside Europe for reasonable shipping cost and shipping time.

Have you tested the imu code both for f130 and f103 ?

**Comment by hoverboardhavoc** [comment]
I've not made the F103 work yet at all. I sent you an email @RoboDurden

**Comment by hoverboardhavoc** [comment]
well it works 🥳  Maybe some tuning required. I want to try on a bigger vehicle

https://github.com/user-attachments/assets/9a0cfdbc-5f63-47f3-b021-18f2e393f75c

**Comment by RoboDurden** [comment]
i don't really know in how the imu is used for balancing. I think the balancing is mainly achieved via the pressure off the foot onto these opto triggered black pads.

**Comment by hoverboardhavoc** [comment]
The two wheels fight with each other so I'm going to run the PID loop on the microcontroller, meaning our requirements are pretty similar @RoboDurden. I'll just send the raw values over the UART bus. I'm currently using `remoteUart`

	typedef struct{
		int16_t     x;
		int16_t     y;
		int16_t     z; 
	} Gyro;

	typedef struct{
		int16_t     x;
		int16_t     y;
		int16_t     z; 
	} Accel;

**Comment by RoboDurden** [comment]
Please link to your github project when the code has been cleaned, so i can see how complex it is.

**Comment by RoboDurden** [comment]
I may want to use one of my gen2.1.3 boards for the ebike.
This is the imu on that board.
Maybe you have some ideas:

![Image](https://github.com/user-attachments/assets/2413f543-aa80-495c-97fa-47fef2138f1d)

**Comment by hoverboardhavoc** [comment]
Maybe it's worth playing with https://github.com/hoverboardhavoc/HoverboardImu It has lots of different branches with different arduino IMU libraries. It will build from platform IO.

    ICM-20948
    ITG3200
    LSM6
    MPU6050
    MPU9250

My IMU gave an unusual WHO_AM_I response (`0x79`) but it still works like a MPU6050

**Comment by RoboDurden** [comment]
> Not sure if it is useful, but I found out that the IMU used with the F130K6/8 is a Bosch BMI120. Not sure how they used it, as it is definitely wired in I2C mode, and those pins do not really support it..

@ThorsAvenger which layout do you mean ?
I already suspect the 2.2 to have a IMU connected via SPI, which the BOSCH BMI120 would support:

```c++
3.2.1 Primary Interface I2C/SPI Protocol Selection
The protocol is automatically selected based on the chip select CSB pin behavior after power-
up.
At reset / power-up, BMI120 is in I2C mode. If CSB is connected to VDDIO during power-up and
not changed the sensor interface works in I2C mode. For using I2C, it is recommended to hard-
wire the CSB line to VDDIO. Since power-on-reset is only executed when, both VDD and VDDIO are
established, there is no risk of incorrect protocol detection due to power-up sequence.
If CSB sees a rising edge after power-up, the BMI120 interface switches to SPI until a reset or
the next power-up occurs. Therefore, a CSB rising edge is needed before starting the SPI communication. Hence, it is recommended to perform a SPI single read access to the
ADDRESS 0x7F before the actual communication in order to use the SPI interface.
If toggling of the CSB bit is not possible without data communication, there is in addition the
spi_en bit in Register (0x70) NV_CONF, which can be used to permanently set the primary
interface to SPI without the need to toggle the CSB pin at every power-up or reset.
```
https://lcsc.com/datasheet/lcsc_datasheet_1912111437_Bosch-Sensortec-BMI120_C437657.pdf

**Comment by ThorsAvenger** [comment]
According to your numbering it is v3=2.2. I got a response for it using the SoftI2C library, combined with DF_Robots BMI160 library, which I think makes use of bit banging. For SPI you need to be able to set the Chip Select pin, but on my board that pin on the imu is hardwired to 3V3. I've also tried to read the NV_CONF register, but is reads all zeros, meaning the spi_on bit has not been set.

**Comment by RoboDurden** [comment]
I have two pairs of 2.1.3 and they nicely small. Would be great if you get the imu to work as i would like to use such a small gen2 board for my ebike project. So the SoftI2C library is not bound to the PB6/PB7 and PB8/PB9 pins ?
In the manual i read that the bosch imu can be configured to be in spi mode without the need put a rising flank on the spi/i2c pin..

Deepseek just helped me with the correct init code for GD32E230 (might be the same for GD32F130):

```c++
void i2c_config(void) 	// Deepseek
{
    rcu_periph_clock_enable(RCU_GPIOB);
    rcu_periph_clock_enable(RCU_I2C0);

	#ifdef PB6PB7	
		// Configure PB6 (SCL) and PB7 (SDA) as AF open-drain
		gpio_mode_set(GPIOB, GPIO_MODE_AF, GPIO_PUPD_PULLUP, GPIO_PIN_6 | GPIO_PIN_7);
		gpio_output_options_set(GPIOB, GPIO_OTYPE_OD, GPIO_OSPEED_50MHZ, GPIO_PIN_6 | GPIO_PIN_7);
		gpio_af_set(GPIOB, GPIO_AF_1, GPIO_PIN_6 | GPIO_PIN_7);  // AF1 for I2C0
	#else
		// Configure PB8 (SCL) and PB9 (SDA) as AF open-drain
		gpio_mode_set(GPIOB, GPIO_MODE_AF, GPIO_PUPD_PULLUP, GPIO_PIN_8 | GPIO_PIN_9);
		gpio_output_options_set(GPIOB, GPIO_OTYPE_OD, GPIO_OSPEED_50MHZ, GPIO_PIN_8 | GPIO_PIN_9);
		gpio_af_set(GPIOB, GPIO_AF_1, GPIO_PIN_8 | GPIO_PIN_9);  // AF1 for I2C0
	#endif	

    // I2C configuration (Standard mode: 100kHz)
    i2c_deinit(I2C0);
    i2c_clock_config(I2C0, 100000, I2C_DTCY_2);
    i2c_mode_addr_config(I2C0, I2C_I2CMODE_ENABLE, I2C_ADDFORMAT_7BITS, 0x00);
    i2c_enable(I2C0);
    i2c_ack_config(I2C0, I2C_ACK_ENABLE);
}	
```

Then on my 2.3.2 the ads1115 on PB6/PB7 was detected, and now there is an i2c device on PB8/PB9: 0x68
i hope, it is a MPU6050

**Comment by RoboDurden** [comment]
[<img width="1315" height="779" alt="Image" src="https://github.com/user-attachments/assets/dbb4ad17-668d-47c9-9482-4f56bbefb89b" />](https://youtu.be/KDhp0SOtsDc)

Ah, now i undestand, accel is indeed the acceleration but it contains the 9.80665 m/s² Gravitation. That is why the xyz values change when i rotate the board.
So i should be able to detect braking. Like this small push in x direction:
<img width="569" height="564" alt="Image" src="https://github.com/user-attachments/assets/835d231f-2f7d-4bfe-863a-62964ba2a107" />
Gravtition here in the y orientiation of the onboard mpu6500.

But the gyro data still seems like rotational speed to me and not rotation acceleration. These are the readings of a 90° left rotation around the x-axsis of the mpu6500:

<img width="820" height="577" alt="Image" src="https://github.com/user-attachments/assets/09ae2a55-021a-4c87-b27f-c4098d7b93f7" />

If it would be acceleration, there should be a positive spike that initiates the rotation followed by a negative spike when the roation is finished.
But gyro.x stays negative for the whole rotation, being minimal in the middle - where the rotation speed is max. Just as i would expect it from rotational speed.
The first image of the translational acceleration instead does show the acceleration and braking with the push in x-direction.


Okay, with the kalman filter, the direction of the gravtition can assist the integration of the rotational acceleration to give a drift free rotation angle. Well only for the x and y rotation. for the z-rotation, the gravitaional vector will not change.  For hoverboard balancing however, you only need the x-rotation :-)
For a lawn mover however, the z-rotation is what we need to change direction :-(
So adding an compass module to the i2c bus might be a good idea.

As i do not need that for my ebike brake/pedaling detection, i am happy with the raw data.

if your  kalman filter indeed removes rotational drift for the x and y-axis, it should be run by the hoverboard. With optional i2c compass module ?
esp32/rpi might be busy with wifi and vision-ai.

Is there some optiocal mouse i2c module that could track the underground of a lawn mover to feed into a kalman filter together with the acceleration data ?



Deepseek, Is there a cheap i2c module that works like an optical mouse to track the underground of lawn mover giving speed and position of the driving robot ?

$9 PMW3901 Optical Flow Sensor Module : https://de.aliexpress.com/item/1005007333832846.html

>Yes, there are affordable I2C modules that function like optical mouse sensors, designed to track ground movement for robotics applications like lawn mowers. Here's a practical solution:
...
Position Drift: Optical flow sensors accumulate error over time.
Solution: Fuse with wheel encoders or IMU (accelerometer/gyro) using sensor fusion (e.g., Kalman filter).

Indeed, combining iOdom (wheel rotation) with the mpu6500 data in a kalman filter might remove the slip of the wheels on the gras :-)

**Comment by RoboDurden** [comment]
You want to emulate the onboard i2c IMU with an EPS32 so that you can trick the boards to spin motors like you want them to do ?
Looks like a lot of work to me and every layout has different IMU.
I am not sure if my imu support for 2.1.20 will work for even a few other boards.

I think the original hoverboard logic also includes the two IR-PHOTO sensors on the backside.
But maybe these only check that the foot is firmly on the hoverboard placed.


But i was thinking if the old firmware could be read to a bin file.
And then simply let Gemini anaylse the assambler code to rework the i2c protocol for the onboard IMU.
Might not even need to recompile into c-code.
I2c communication is very simple and most imu have 0x68 as the i2c address.
So it could be quite simple for ai to find the read/write operations to 0x68..

Good use case for ai.

**Comment by RoboDurden** [comment]
sniffing with such a cheap logic analyzer https://de.aliexpress.com/item/1005009676857208.html ?
Or with ESP32 as I2C slave and somehow outputing the sda stream ?

What layout are you working with ?
Why are you not happy with my firmware ?
Have you tried my new 	`#define BLDC_SINE_BOOSTER		// can boost speed by 15% starting from 87% throttle.` ?

But yes, these GD32E230 boards are so hard to flash, removing the imu and control the motors like these hovercarts, might indeed work.

<img width="540" height="423" alt="Image" src="https://github.com/user-attachments/assets/5aa3ad07-8c69-4ef9-8632-6239c251e584" />

**Comment by WestlingPi** [comment]
I have started working on getting the IMU working on my 2.2.3 board. I have managed to get the i2c initialization code to work on STM32F103. On 2.2.3 board, i2c is connected to PB8/PB9 so it requires remapping of I2C0 to PB8/PB9. I only added STM32F103 code for PB8/PB9. Do you want me to add untested STM32F103 code for PB6/PB7? (I think it is same, but without the remapping). With this I managed to find a i2c device on device address 0x68 with i2c_scanner(). With dump_i2c_registers() I got 0x72 from the WHO_AM_I register (0x75). According to Gemini, it is a non-standard value probably from a MPU6500 or ICM20600 which is 99.9 compatible with MPU6050 (the difference is how temperature sensor is interpreted). When the board starts up, the IMU on this board is in sleep mode (PWR_MGMT_1 register (0x6B) is 64) and all sensor registers reads as 0. But the mpu_config() takes it out of sleep mode (PWR_MGMT_1 set to 0x01). After that all the sensor registers is working. (According to Gemini, 10ms sleep after switching to PLL clock is a bit risky so I changed it to 50ms as recommended by Gemini).

Which version of the MPU6050 code should I use? I see there are 2 versions: mpu6050.c (#ifdef MPU_6050old) and imuMPU6050.c (#ifdef MPU_6050). I have used and updated mpu6050.c because MPU_6050old was the only one I found used in any of the board files (defines_2-1-20.h use MPU_6050old).

I have tested this on my 2.2.3 board. I have ONLY tested I2C_Init(), i2c_scanner(), dump_i2c_registers() and MPU_Init() called from main, and it is working. Next step to add support for IMU in RemoteROS2 and send the raw sensor readings from the hoverboard to ROS2 controller so that I can use the sensor fusion in ROS2. But maybe you prefer a smaller pull request first with just the i2c stuff for STM32F103 and minor fixes to mpu6050.c? (And also it will probably be until after new years until I get time from the ROS2 stuff).

https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/pull/19

**Comment by RoboDurden** [comment]
Good to hear that you had success with another board. Even another target.
Looking at my PilotHoverbike code for 2.1.20 i think i also used the old code:

```c++
//#define MPU_6050		// Gemini tried for clones..
#define MPU_6050old	
#define I2C_PB6PB7	// IMU
```

And imuMPU6050.c was just a Gemeni try that failed.

Because i had problems with other boards to wake up the imu.
So if you found new code to make your imu work, this might help to support more boards.

I don't really care for your ROS2 stuff, as long as you stay in your Remote and Pilot files, as i have no time to test ROS2 anyway.
That is why i made this RemoteXY and PilotXY structure.
So when you think your ROS2 is safe for others, i will accept any pull request - as long as you do not add changes to main.c , etc. which i can not for sure consider 100% safe.
And to understand your changes to i2c/imu files, it would help me to see your ROS stuff make use of it.

Try to stay consistent to the way i added imu data to RemoteUart.c 
So also using the define `SEND_IMU_DATA`

Thank you for your ongoing contributions and taking care of the ROS2 feature.

**Comment by WestlingPi** [comment]
Ok, then I will continue to use mpu6050.c (MPU_6050old). It seems to work fine for me so I dont think I need to change it.

I plan to do all the rest of the updates needed in RemoteROS2. I have seen the SEND_IMU_DATA define used in other remotes so I plan to follow that pattern. But I need to read up a little on IMUs and ROS2 drivers for IMUs to know exactly what I need to do, but I think the structure with Remotes (and Pilots) will work nicely. My plan is to fuse the sensor data from the 2 IMUs with wheel odometry to improve localization (wheel slipage etc). Maybe also add a GPS and compass later (not related to this firmware)...

Thanks for accepting the pull request and Merry Christmas!

**Comment by WestlingPi** [comment]
I have done some more research (i.e. AI chatting) and I think Gemini have convinced me that it is a MPU6522 IMU on my 2.2.3 hoverboard. According to Gemini MPU6522 is a OEM version of MPU6500 with same physical package as MPU6050 (4x4mm compared to 3x3mm for MPU6500) and reading 0x72 from WHO_AM_I register, otherwise same chip as MPU6500. The good news is that MPU6500/6522 is 99% compatible with MPU6050 so I only need a minor extension of the mpu6050.c code.
For the initialization code, MPU6500 use separate configuration for gyro and accelerometer digital low-pass filter (register 26 and 29), whereas MPU6050 use same configuration for both (register 26). I have added some code to do the extra configuration needed for MPU6500, see https://github.com/WestlingPi/Hoverboard-Firmware-Hack-Gen2.x-GD32/commit/a3a91223493600b2ae447f27e08479167c7e089b. However, I dont have access to my hoverboard right now and I want to do some more testing before I submit a pull request.

Another thing that I would like you feedback on is the lowpass filter at the end of MPU_ReadAll() https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/blob/main/HoverBoardGigaDevice/Src/mpu6050.c#L362 . In RemoteROS2, I want to send the raw sensor data to ROS2 with as little delay and filtering as possible and feed it into a Kalman filter (in robot_localization package). So I would like to move the lowpass filter in MPU_ReadAll() to your Pilot/Remote. If it is needed by your pilot, I think it is better to have it there instead of the sensor code that is used by all Pilots/Remotes. If it is ok with you, I can move it when I start with my RemoteROS2 updates.

**Comment by RoboDurden** [comment]
Couldn't you replace the 4 in DoLowPass with a define in the .h file ?
Like 
```c++
#ifndef IMU_LP 
  #define IMU_LP 4
#endif
```

And in .c file

```c++
#if IMU_LP > 0
  DoLowPass (IMU_LP,...)
#endif
```

I would like to keep the lowpass so that other pilotxy and remotexy can make use of that feature.

**Comment by WestlingPi** [comment]
Ok, sure. A define would be fine for me too. I can add it when I start on the RemoteROS2 changes.

Are you aware that there is a lowpass filter in the IMU hardware (configured by register 26 for MPU6050 and 26 & 29 for MPU6500)? Maybe you can achieve the same thing with a higher(?) value of register 26/29 and thus would not need an additional lowpass filter in the firmware software? Just a thought that might be worth exploring... Otherwise, I think the current configuration and a define for the software lowpass filter will work for me.

**Comment by WestlingPi** [comment]
Now I have tested the MPU6500 initialization on my 2.2.3 board and submitted a pull request: https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/pull/20

I am also testing an update of the RemoteROS2 that will send the IMU sensor values to ROS2. I will submit another PR when I have tested it some more.

**Comment by WestlingPi** [comment]
I have been working on sending IMU data from the RemoteROS2 (and I think it is ready now). However, when testing RemoteROS2 I ran in to some problems that I think is i2c related.

The symptom I get is that reading the sensor values works great for a while, then suddenly it stops working and never recovers. At first I thought it was just after a while (10-30s after power on) but later I discovered that it is probably related to when increasing the bldc speed from 0.
@RoboDurden I have seen some comments in the code that maybe you have experienced similar issues, eg:
`I2C_Init();	// try to re-init if i2c bus fails. Need when I2C_SPEED is 400000 instead of 200000`

So the last few days I have been debugging this issue. I added some debug counters to i2c_readBytes(), one for every error case, and it turns out that when it fails for me, it is always in this part of i2c_readBytes():

```
    i2c_master_addressing(i2c_periph, devAddr << 1, I2C_TRANSMITTER);
    tmo = 0;
    while (!i2c_flag_get(i2c_periph, I2C_FLAG_ADDSEND)) {
        if (++tmo > I2C_TIMEOUT) {
            i2c_stop_on_bus(i2c_periph);
            i2cE3++;
            i2c_hard_reset(i2c_periph);
            return I2C_ERR;
        }
    }
    i2c_flag_clear(i2c_periph, I2C_FLAG_ADDSEND);
```

At first I tried to toggle the devAddr between 0x68 and 0x69 on every failed attempt, but that didnt help. (There is an AD0 pin on mpu6050/6500 that selects i2c address 0x68 or 0x69 that I thought it could be something wrong with - but the pcb seems ok).

Then Gemini suggested to add a i2c_hard_reset():
```
void i2c_hard_reset(uint32_t i2c_periph) {
	// 1. Force the I2C peripheral into reset state
	i2c_software_reset_config(i2c_periph, I2C_SRESET_SET);
	for(volatile int d=0; d<1000; d++);
	i2c_software_reset_config(i2c_periph, I2C_SRESET_RESET);

	// 2. Re-initialize the I2C peripheral from scratch
	I2C_Init();
}
```

With this code, whenever i2c_readBytes() fail, it recovers and is able to read the next time. According to Gemini the i2c bus is sensitive to electromagnetic noise from the bldc driver and in particular the STM32F103 hardware implementation of i2c is more sensitive than the corresponding GD32 hardware.

@RoboDurden Do you think I should add the i2c_hard_reset() to i2c_readBytes(), or do you have any other idea? With this fix I am able to run the motor (accelerate, decelerate, continuous  speed etc). The i2c reads fails sometimes, but it always recovers. I dont experience any "hangings" anymore. I have not tried yet if it is enough with only I2C_Init() to recover. Also, do you prefer to have the recovery in i2c_readBytes() or in the users (in your case PilotHoverbike and my case RemoteROS2)?

**Comment by RoboDurden** [comment]
Oh yes, great you solved this issue! I got quite frustrated when my hoverBike imu reading always failed.
So yes, your hard_reset should go into readBytes. 
But not always hardReset but only when readBytes fail, do a hardReset and repeat readBytes ? Maybe a short delay is then needed ?

**Comment by WestlingPi** [comment]
Yes, only hard reset when readBytes fails. I dont think a short delay is needed because in main, i2c_scanner() is called right after I2C_Init().

I can do some more testing and then submit a PR together with the RemoteROS2 code.

**Comment by WestlingPi** [comment]
I have done some more testing. In my RemoteROS2 i try to read IMU sensors one more time immediate after a failure. 

```
#ifdef SEND_IMU_DATA
	if (MPU_ReadAll() == SUCCESS)
	{
		i2cNoError++;
		//Send IMU data to ROS2
		// ...
	}
	else {
		i2cError++;
			if (MPU_ReadAll() == SUCCESS) {
				i2cNoError2++;
			}
			else {
				i2cError2++;
			}
	}
#endif
```

And i2c_readBytes contains the hard reset recovery as previous post. After about 3 minutes of "mixed" driving, I get the following debug counter:

<img width="350" height="185" alt="Image" src="https://github.com/user-attachments/assets/6cee0d39-1011-4656-ba36-bbb13c5f184a" />

I.e. after about 16000 read attempts, it has failed 44 times. All of the failures happened at the 3rd error case in i2c_readBytes (i2cE3 - see previous post), and an immediate read attempt after the hard reset recover works after all 44 failures (i2cNoError2=44 and i2cError2=0).

So I think it hard reset works well and no need for delay. I will clean up the code and submit a PR (probably another day).

**Comment by WestlingPi** [comment]
Pull request submitted: https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/pull/21
